src/DynamoCore/Graph/Nodes/CodeBlockNode.cs

```diff
@@ -627,30 +643,122 @@ internal void ProcessCodeDirect()
             OnNodeModified();
         }
 
+        /// <summary>
+        /// Undefine a function definition in a code block node if it has any.
+        /// </summary>
+        internal void UndefineFunctionDefinitions()
+        {
+            var funcDefs = ParseParam.ParsedNodes.OfType<FunctionDefinitionNode>();
+            foreach (var funcDef in funcDefs)
+            {
+                libraryServices.LibraryManagementCore.SetFunctionInactive(funcDef);
+            }
+        }
+
+        /// <summary>
+        /// This method skips parsing and only recompiles a CBN in a second pass after all other function definitions
+        /// have been compiled. It re-uses the ParseParam (ASTs etc.) of the CBN compiled from the first pass.
+        /// </summary>
+        /// <param name="errorMessage"></param>
+        /// <param name="warningMessage"></param>
+        internal void RecompileCodeBlockAST(ref string errorMessage, ref string warningMessage)
+        {
+            BuildStatus buildStatus = null;
+            try
+            {
+                int blockId = Constants.kInvalidIndex;
+                var core = libraryServices.LibraryManagementCore;
+
+                bool parsingPreloadFlag = core.IsParsingPreloadedAssembly;
+                bool parsingCbnFlag = core.IsParsingPreloadedAssembly;
+
+                core.IsParsingPreloadedAssembly = false;
+                core.IsParsingCodeBlockNode = true;
+                // This is always the second pass of precompiling a CBN so we need to reset this flag to false here.
+                core.IsCodeBlockNodeFirstPass = false;
+
+                core.ResetForPrecompilation();
+
+                // FunctionDefinitonNodes have already been compiled.
+                // Recompile the rest of the ASTs with function definitions in place.
+                var astNodes = ParseParam.ParsedNodes.Where(x => !(x is FunctionDefinitionNode));
+
+                // Clone a disposable copy of AST nodes for PreCompile() as Codegen mutates AST's
+                // while performing SSA transforms and we want to keep the original AST's
+                var codeblock = new CodeBlockNode();
+                var nodes = astNodes.Select(NodeUtils.Clone).ToList();
+                codeblock.Body.AddRange(nodes);
+
+                buildStatus = CompilerUtils.PreCompile(string.Empty, core, codeblock, out blockId);
+
+                core.IsParsingCodeBlockNode = parsingCbnFlag;
+                core.IsParsingPreloadedAssembly = parsingPreloadFlag;
+
+                ParseParam.AppendErrors(buildStatus.Errors);
+                ParseParam.AppendWarnings(buildStatus.Warnings);
+            }
+            catch (Exception)
+            {
+                buildStatus = null;
```

Comment:
should we log something?

Reply:
I'm not sure if we can log any useful information here that would let the user debug the node. For example, how would we identify this code block node for the user? We are catching a similar exception [here](https://github.com/DynamoDS/Dynamo/blob/3df0eede2b94dc967354535b35703d9d13c0dd0f/src/Engine/ProtoCore/Utils/CompilerUtils.cs#L371) and not logging anything there either.

Any compilation warnings should appear on the code block node in any case.